src/Database/Table/ColumnAccessCache.php

```diff
@@ -0,0 +1,195 @@
+	protected $selection;
+
+	/** @var Cache */
+	protected $cache;
+
+	/** @var string */
+	protected $generalCacheKey;
+
+	/** @var string */
+	protected $specificCacheKey;
+
+	/** @var array of touched columns */
+	protected $accessedColumns;
+
+	/** @var array of earlier touched columns */
+	protected $previousAccessedColumns;
+
+	/** @var bool should instance observe accessed columns caching */
+	protected $observeCache = false;
+
+
+	public function __construct(Selection $selection, IStorage $cacheStorage = null)
+	{
+		$this->selection = $selection;
+		$this->cache = $cacheStorage ? new Cache($cacheStorage, 'Nette.Database.' . md5($selection->getConnection()->getDsn())) : null;
+	}
+
+
+	public function getStorage(): ?IStorage
+	{
+		return $this->cache ? $this->cache->getStorage() : null;
+	}
+
+
+	/**
+	 * Returns general cache key independent on query parameters or sql limit
+	 * Used e.g. for previously accessed columns caching
+	 */
+	public function getGeneralCacheKey(): string
+	{
+		if ($this->generalCacheKey) {
+			return $this->generalCacheKey;
+		}
+
+		$key = [__CLASS__, $this->selection->getName(), $this->selection->getSqlBuilder()->getConditions()];
+		$trace = [];
+		foreach (debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS) as $item) {
+			$trace[] = isset($item['file'], $item['line']) ? $item['file'] . $item['line'] : null;
+		}
+
+		$key[] = $trace;
+		return $this->generalCacheKey = md5(serialize($key));
+	}
+
+
+	public function setGeneralCacheKey(?string $key): void
+	{
+		$this->generalCacheKey = $key;
+	}
+
+
+	/**
+	 * Returns object specific cache key dependent on query parameters
+	 * Used e.g. for reference memory caching
+	 */
+	public function getSpecificCacheKey(): string
+	{
+		if ($this->specificCacheKey) {
+			return $this->specificCacheKey;
+		}
+
+		return $this->specificCacheKey = $this->selection->getSqlBuilder()->getSelectQueryHash($this->getPreviousAccessedColumns());
+	}
+
+
+	public function setSpecificCacheKey(?string $key): void
+	{
+		$this->specificCacheKey = $key;
+	}
+
+
+	/**
+	 * @return mixed
+	 */
+	public function getAccessedColumns()
+	{
+		return $this->accessedColumns;
+	}
+
+
+	/**
+	 * @param mixed
+	 */
+	public function setAccessedColumns($accessedColumns): void
+	{
+		$this->accessedColumns = $accessedColumns;
+	}
+
+
+	public function setAccessedColumn(string $key, bool $value): void
+	{
+		$this->accessedColumns[$key] = $value;
+	}
+
+
+	/**
+	 * Loads cache of previous accessed columns and returns it.
+	 * @return array|bool
+	 */
+	public function getPreviousAccessedColumns()
+	{
+		if ($this->cache && $this->previousAccessedColumns === null) {
+			$this->accessedColumns = $this->previousAccessedColumns = $this->cache->load($this->getGeneralCacheKey());
+			if ($this->previousAccessedColumns === null) {
+				$this->previousAccessedColumns = [];
+			}
+		}
+
+		return array_keys(array_filter((array) $this->previousAccessedColumns));
+	}
+
+
+	/**
+	 * @param mixed
```

Comment:
$previousAccessedColumns  has typehint array and here is mixed

Reply:
My bad. I copied properties as they were. Fixed on all properties.